### PR TITLE
[TIKA-3353] Prometheus and JMX monitoring over micrometer

### PR DESCRIPTION
**Pull request by Subhajitdas298**
New standard metrics collection system to be implemented, with help Micrometer metrics system, with CXF pluggability.

The system is extensible and can be extended t support other monitoring system later on.

This is for Branch 1x

### CODE REVIEW

**Comment by Subhajitdas298** [comment]
**Please note**: There are new dependencies added for this purpose.

**Comment by Subhajitdas298** [comment]
@tballison Please review the PR and let me know, if anything has to be modified.

**Comment by tballison** [comment]
@Subhajitdas298 , let's hold off on this for main/2.x for now.  I'd like to upgrade to log4j2, and I'd like to make these extra dependencies loaded only if the user selects "monitor"...prob next week?

Thank you for your work on this!

File: tika-server/src/main/java/org/apache/tika/server/metrics/ServerStatusMetrics.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * Binds server status metrics to meter registry.
+     * @param meterRegistry the meter registry to bind to.
+     */
+    @Override
+    public void bindTo(@NotNull MeterRegistry meterRegistry) {
+        Gauge.builder("server.status.lastparsed", serverStatus, ServerStatus::getMillisSinceLastParseStarted)
+                .description("Last parsed in milliseconds")
+                .register(meterRegistry);
+        Gauge.builder("server.status.restarts", serverStatus, ServerStatus::getNumRestarts)
+                .description("Last parsed in milliseconds")
```
Comment (lewismc) [inline tika-server/src/main/java/org/apache/tika/server/metrics/ServerStatusMetrics.java:54]: The help here is incorrect.

File: tika-server/src/main/java/org/apache/tika/server/metrics/ServerStatusMetrics.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                .register(meterRegistry);
+        Gauge.builder("server.status.files", serverStatus, ServerStatus::getFilesProcessed)
+                .description("Last parsed in milliseconds")
```
Comment (lewismc) [inline tika-server/src/main/java/org/apache/tika/server/metrics/ServerStatusMetrics.java:57]: The help here is incorrect.

**Review by lewismc — COMMENTED** [review]
@Subhajitdas298 I was talking to @tballison and discpvered that this was merged into 1.X. I would like to get this into tika-server 2.X.
I have a couple of questions
1. Why did you pick Prometheus over say StatsD?
2. Do you have any documentation which lays out the generated metrics? I made some comments on Guage's which you created... is there anything else?
3. It occurs to me that turning on Prometheus metrics is driven by a command line flag... is that true? If so, it would be great it we could extend that configuration to for example the downstream binaries such as the tika-docker and tika-helm.

Thanks for any information.

**Comment by Opa-** [comment]
@tballison would you know if this is still planned to be merged into 2.x ?

Or maybe there's another feature covering this for 2.x that I missed.

**Comment by tballison** [comment]
Frankly, at this point, we should target 3.x. Are you currently using this functionality in 1.x? I don't know enough about this to know if it is general enough and suited for continued support. @lewismc would you have any time to help with this?

**Comment by Opa-** [comment]
I'm in the process of migrating from 1.x to 2.x and just discovered this 1.x feature. Don't bother adding it to 2.x just for me, the `/status` route cover my use case for now.
Although these kind of metrics would be neat to have in 3.x :)

**Comment by lewismc** [comment]
Hi @tballison 

> ...would you have any time to help with this?

Yes once I’ve finished an upgrade to JUnit Jupiter for Nutch. This would be a great feature for Tika 3.X server.

**Comment by tballison** [comment]
See recent work on the 4.x/main branch. wdyt?

https://github.com/apache/tika/pull/3064

**Comment by andygrunwald** [comment]
@tballison Looks pretty good at its first glance! Thanks
Awaiting a 4.x release with this feature to properly test it.

**Pull request merged**